### ISSUE TITLE
Block signatures as a source of randomness

### ISSUE THREAD

**Issue by AnnaShaleva** [bane-labs/go-ethereum#119]
# Rationale

Since Eth post-merge (https://eips.ethereum.org/EIPS/eip-4399) block's `mixHash` field stores the outpur of RANDAO and is used as the source of randomness in `PREVRANDAO` opcode. In NeoX `mixHash` block field is used to store NextConsensus related information, and thus, it can't be used as a reliable source of randomness (its value isn't being changed often). We need a trusted source of randomness to base `PREVRANDAO` opcode on.

# Implementation

Once #48 is implemented and BLS signatures are successfully adopted, we suggest to use block's signatures (constant for the blocks of the same height!) as a trusted source of randomness. We need to override the `PREVRANDAO` opcode to refer to block's signatures instead of `mixHash`.

**Comment by vncoelho** [comment]
After phase 3 of dbft with antimev we can use the hash of the decrypted txs as a nonce.
Some adjustments can be made for a deterministic calculation.

This will be simple embedded in the current design.

**Comment by roman-khimov** [comment]
You don't always have encrypted transactions, while you always have some block signature. Technically, given that this signature is a part of a block and is being hashed to get the block hash, with #48 block hash itself is random enough. Still, `PREVRANDAO` has to have some meaning as well.

**Comment by vncoelho** [comment]
> You don't always have encrypted transactions, while you always have some block signature. Technically, given that this signature is a part of a block and is being hashed to get the block hash, with #48 block hash itself is random enough. Still, `PREVRANDAO` has to have some meaning as well.

It is true that we may not always have encrypted txs.
But we will have the mechanism for the thresholds aggregation based on BLS, we can just sign something and use the aggregated signatures (the primary can do this simple nounce encrypted in such cases - not strong as the other one but still not seen by the other CNs backups).
As I said on my previous comment, it is not just the hash of the encrypted, but some deterministic calculation based on the result of the aggregation of the BLS fragments, normally, these fragments will be the encrypted of AES keys made by the users.

**Comment by dusmart** [comment]
> You don't always have encrypted transactions, while you always have some block signature.

From a user's perspective, they cloud make sure that their `PREVRANDAO` calling is contained in a encrypted transaction. And of cause they could give up the ability of adding random source to the `PREVRANDAO` by using normal transactions. While current `blockhash` aka `blockhash(block.number)` is not available for them in the transaction.

> As I said on my previous comment, it is not just the hash of the encrypted, but some deterministic calculation based on the result of the aggregation of the BLS fragments, normally, these fragments will be the encrypted of AES keys made by the users.

Given that this `PREVRANDAO` opcode (previous `DIFFICULTY`) is usually considered as a randomness source. Users' secret input such as AES keys are a good fact for building a fair random value.

Although current blockhash and signature is a good randomness source, they could not be determined when calculating the transaction receipt which are then used to generate the block's state root.

Maybe we could build that value from some or all of the AES keys and a base seed from all of the CNs (such as a BLS signature of the current block number in dBFT phase two).

What's more, I think it's the dAPP's duty to reminder users to use the encrypted transaction to add randomness source if they choose to building the randomness only upon the `PREVRANDAO`.

**Comment by roman-khimov** [comment]
> Although current blockhash and signature is a good randomness source, they could not be determined when calculating the transaction receipt which are then used to generate the block's state root.

True. But given the fact that we gonna have an additional dBFT round, we can reuse some material from the penultimate one, it's just a matter of saving it then.

**Comment by dusmart** [comment]
> True. But given the fact that we gonna have an additional dBFT round, we can reuse some material from the penultimate one, it's just a matter of saving it then.

You're right. Actually the state root can not be determined until all of the 
encrypted transactions are unveiled. We could reuse some material from the penultimate one as `PREVRANDAO` and then calculate transaction receipt for all.

The only problem left is **should we add randomness from encrypted transactions or not**. If we add them, the randomness will be more fair while it indeed needs some additional calculation for all CNs and it's not stable because not all transactions are going with an encrypted transaction. I'm OK with both of them.

**Comment by roman-khimov** [comment]
I think we can have random enough BLS signature, why do we need anything additional from encrypted transactions?
